tst.c

```diff
@@ -189,31 +203,45 @@ void *tst_ins_del(tst_node **root, const char *s, const int del, const int cpy)
     if (strlen(s) + 1 > STKMAX / 2) /* limit length to 1/2 STKMAX */
         return NULL;                /* 128 char word length is plenty */
 
-    pcurr = root;                     /* start at root */
-    while ((curr = *pcurr)) {         /* iterate to insertion node  */
-        diff = *p - curr->key;        /* get ASCII diff for >, <, = */
-        if (diff == 0) {              /* if char equal to node->key */
-            if (*p++ == 0) {          /* check if word is duplicate */
-                if (del) {            /* delete instead of insert   */
-                    (curr->refcnt)--; /* decrement reference count  */
-                    /* chk refcnt, del 's', return NULL on successful del */
-                    return tst_del_word(root, curr, &stk, cpy);
-                } else
-                    curr->refcnt++; /* increment refcnt if word exists */
-                return (void *) curr->eqkid; /* pointer to word / NULL on del */
-            }
-            pcurr = &(curr->eqkid); /* get next eqkid pointer address */
-        } else if (diff < 0) {      /* if char less than node->key */
-            pcurr = &(curr->lokid); /* get next lokid pointer address */
-        } else {                    /* if char greater than node->key */
-            pcurr = &(curr->hikid); /* get next hikid pointer address */
-        }
-        if (del)
-            tst_stack_push(&stk, curr); /* push node on stack for del */
+    pcurr = root;
+    while ((curr = *pcurr) && *p) {
```

Comment:
The logic of this while loop is not as same as the original one! What if `curr != NULL` and also `*p == 0` are both True after going out of the while loop? Thinking carefully about under what situatuin should `tst_del_word` be called.

So also the part of `tst_ins` should be changed.